### DISCUSSION TITLE
SvelteFlow Dynamic EdgeMarkers

### DISCUSSION THREAD

**Discussion by Everduin94** [xyflow/xyflow#3990] [Q&A] [answered]
Hello! First I wanted to say this is a really cool project and I've enjoyed working with it and seeing what it can do.

My question is about adding a basic MarkerType.Arrow to a custom edge via the `markerEnd` prop.

I have read the documentation
- Custom Edges: https://svelteflow.dev/examples/edges/custom-edge
- As well as the note here: https://svelteflow.dev/api-reference/components/base-edge#notes

However, (and I think a lot of the SvelteFlow documentation suffers from this), there aren't any examples of *dynamically creating edges with a markerEnd*. - i.e. When a user creates a new edge.

For example, in the Custom Edges example. All of the data is hard coded in a ts file. If you delete and re-apply the edge, the `MarkerType` is lost.

I have something like this. However, I don't see how I can set this `markerEnd` prop on the `edges` store when a user creates a new edge. - If I hard code the component prop to `MarkerType.Arrow` that doesn't work either. AFAIK, it's actually supposed to be some `url(#id)` value. But, I'm not sure what the url value is or how it's determined.

```html
<script>
// .... custom edge props
export let markerEnd // <-- Always undefined on new edge
</script>

<BaseEdge path={edgePath} {markerEnd} />
<EdgeLabelRenderer>
  <!-- Custom stuff -->
</EdgeLabelRenderer>
```

**Comment by Everduin94** [comment, marked as answer]
For my use-case I believe I found an answer. - Passing `defaultEdgeOptions={{markerEnd: { type: MarkerType.Arrow }}}` to `SvelteFlow`.

**Comment by Nemerion** [comment]
Hi, I am having the same issue regarding the creation of custom markers, because for example, the documentations says that I need to pass a `string` to the `markerEnd` property, but in the example they pass an `object`. So then I look forward and put a console log to see how Flow manages this and it shows a `url("type=arrow")` function that DOES NOT appear in the doc nor the code, so I cannot use it anyway.

The docs also say to use a `EdgeMarker` type, but this is simply not truth, because the moment I pass it it complains, as it only takes a `string`

**Reply by bcakmakoglu** [reply]
You're confusing `edge.markerEnd` (which allows an `EdgeMarker` type object) and `EdgeProps.markerEnd`.

The way it works is this:
React/SvelteFlow will read the markers from your edges, create a marker for each unique marker than can be found and then pass the *reference to that marker as a prop to your edges* (`EdgeProps.markerEnd`).

You cannot pass an `EdgeMarker` type object to `<BaseEdge>` - it expects a `string` that represents the *reference* to the marker (`url(...)`).

**Reply by bcakmakoglu** [reply]
For reference:
[`edge.markerEnd`](https://svelteflow.dev/api-reference/types/edge#marker-end)
```ts
const edge = {
	id: 'e1->2',
	source: '1',
	target: '2',
	markerEnd: {
		type: MarkerType.Arrow,
		color: 'hotpink',
	}
}
```

[`EdgeProps.markerEnd`](https://svelteflow.dev/api-reference/types/edge-props#marker-end)
```ts
const CustomEdge = (props: EdgeProps) => {
	console.log(props.markerEnd) // logs: `url("type=arrow&color=hotpink")`
}
```

**Reply by bcakmakoglu** [reply]
Also, check the [`Edge Markers`](https://svelteflow.dev/examples/edges/edge-markers) example which demonstrates how to use a custom marker.

**Reply by Nemerion** [reply]
Thanks for the response, but my problem is: how can I modify those props INSIDE my custom edge? Because I can do it from the outside with the `EdgeMarker` object type when I build it, but on the Custom Edge component that is a string with a `url(...)` format, which I tried to manipulate directly, but with no success...
For example, inside my CustomEdge.svelte file I have: `<BaseEdge {path} {markerEnd} {style} {label} />` which takes a `markerEnd: string` with this `url(...)` format. If I pass somethis that is not what the Edge is receiving from the outside as props, it does not renders it at all.

**Reply by Nemerion** [reply]
I specifically want to do this because I don't want to pass the `markerEnd ` object every time I have to render an Edge with the same props, and instead put that inside the Edge component that I am literally building just for that case.

**Reply by bcakmakoglu** [reply]
> I don't want to pass the markerEnd object every time I have to render an Edge 

That's what [`defaultEdgeOptions`](https://svelteflow.dev/api-reference/types/default-edge-options) are for.

**Reply by bcakmakoglu** [reply]
> how can I modify those props INSIDE my custom edge

Same way you would modify any other property of edges, there's nothing special about markers here.

> which I tried to manipulate directly, but with no success...

You're not supposed to manipulate the `url` string given to you. Again, the url string is a *reference to an existing marker*.
Modifying it will not change anything about the markers but will only create a reference to a marker that doesn't even exist.

**Reply by bcakmakoglu** [reply]
[Also read this response I gave in another thread about custom edge markers.](https://github.com/xyflow/xyflow/discussions/4300#discussioncomment-9524224)
